src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs

```diff
@@ -746,18 +725,136 @@ public virtual TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext)
+            var fullName = GetFullHtmlFieldName(viewContext, expression);
+            if (string.IsNullOrEmpty(fullName))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(expression));
+            }
+
+            var type = allowMultiple ? typeof(string[]) : typeof(string);
+            var rawValue = GetModelStateValue(viewContext, fullName, type);
+
+            // If ModelState did not contain a current value, fall back to ViewData- or ModelExplorer-supplied value.
+            if (rawValue == null)
+            {
+                if (modelExplorer == null)
+                {
+                    // Html.DropDownList() and Html.ListBox() helper case.
+                    rawValue = viewContext.ViewData.Eval(expression);
+                    if (rawValue is IEnumerable<SelectListItem>)
+                    {
+                        // This ViewData item contains the fallback selectList collection for GenerateSelect().
+                        // Do not try to use this collection.
+                        rawValue = null;
+                    }
+                }
+                else
+                {
+                    // <select/>, Html.DropDownListFor() and Html.ListBoxFor() helper case. Do not use ViewData.
+                    rawValue = modelExplorer.Model;
+                }
+            }
+
+            if (rawValue == null)
+            {
+                return null;
+            }
+
+            // Convert raw value to a collection.
+            IEnumerable rawValues;
+            if (allowMultiple)
+            {
+                rawValues = rawValue as IEnumerable;
+                if (rawValues == null || rawValues is string)
+                {
+                    throw new InvalidOperationException(
+                        Resources.FormatHtmlHelper_SelectExpressionNotEnumerable("expression"));
+                }
+            }
+            else
+            {
+                rawValues = new[] { rawValue };
+            }
+
+            modelExplorer = modelExplorer ??
+                ExpressionMetadataProvider.FromStringExpression(expression, viewContext.ViewData, _metadataProvider);
+            var isEnum = modelExplorer.Metadata.IsEnum;
+            var innerType =
+                Nullable.GetUnderlyingType(modelExplorer.Metadata.ModelType) ?? modelExplorer.Metadata.ModelType;
+
+            // Convert raw value collection to strings.
+            var currentValues = new HashSet<string>(StringComparer.Ordinal);
+            foreach (var value in rawValues)
+            {
+                if (value == null)
+                {
+                    continue;
+                }
+
+                if (!isEnum)
+                {
+                    // Add original or converted string.
+                    var stringValue = (value as string) ?? Convert.ToString(value, CultureInfo.CurrentCulture);
+                    if (!string.IsNullOrEmpty(stringValue))
```

Comment:
- turns out this special case doesn't exist in MVC 5.  checking on the one at line 808.
- fyi `string.Empty` will result in `<option value=""></option>` and should round-trip as `null` instead unless `ConvertEmptyStringToNull` is `false` (an unusual case).  similarly `null` can round-trip as `string.Empty` in some cases.